### PR TITLE
Use columnar processing in projection operator without filter

### PR DESCRIPTION
**Pull request by sopel39**
When there is no filter condition in `FilterAndProjectOperator` it is
optimal to use columnar version of `PageProcessor` so that for
identity expressions no data is copied. Additionally, this
optimizes memory usage when `FilterAndProjectOperator` output
is used in `InnerJoin` as build table.

FYI: @dain @martint

### CODE REVIEW

**Comment by sopel39** [comment]
This particularly improves memory usage for joins when there is pipeline: ExchangeOperator -> Projection -> HashBuilderOperator.

For example, query TPCH.Q5:
``` SELECT n.name, sum(l.extendedprice * (1 - l.discount)) AS revenue FROM customer c, orders o, lineitem l, supplier s, nation n, region r WHERE c.custkey = o.custkey AND l.orderkey = o.orderkey AND l.suppkey = s.suppkey AND c.nationkey = s.nationkey AND s.nationkey = n.nationkey AND n.regionkey = r.regionkey AND r.name = 'ASIA' AND o.orderdate >= DATE '1994-01-01' AND o.orderdate < DATE '1994-01-01' + INTERVAL '1' YEAR GROUP BY n.name ORDER BY revenue DESC;```

Explain (part):
```
 Fragment 5 [HASH]                                                                                                                                                                                          
     Output layout: [nationkey, suppkey, extendedprice, discount, nationkey, suppkey, $hashvalue_115, $hashvalue_115]                                                                                       
     Output partitioning: HASH [nationkey, suppkey]                                                                                                                                                         
     - Project => [extendedprice:double, $hashvalue_115:bigint, nationkey:bigint, discount:double, suppkey:bigint]                                                                                          
             $hashvalue_115 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("nationkey"), 0)), COALESCE("$operator$hash_code"("suppkey"), 0))                                   
         - InnerJoin[("orderkey_46" = "orderkey_47")] => [orderkey:bigint, nationkey:bigint, orderkey_46:bigint, $hashvalue_103:bigint, orderkey_3:bigint, extendedprice:double, discount:double, suppkey:bi
             - Project => [orderkey:bigint, nationkey:bigint, orderkey_46:bigint, $hashvalue_103:bigint]                                                                                                    
                 - RemoteSource[6] => [orderkey:bigint, nationkey:bigint, orderkey_46:bigint, $hashvalue_103:bigint, $hashvalue_104:bigint]                                                                 
             - Project => [orderkey_3:bigint, extendedprice:double, discount:double, suppkey:bigint, orderkey_47:bigint, $hashvalue_112:bigint]                                                             
                 - RemoteSource[9] => [orderkey_3:bigint, extendedprice:double, discount:double, suppkey:bigint, orderkey_47:bigint, $hashvalue_112:bigint, $hashvalue_113:bigint]  
...
 Fragment 9 [SOURCE]                                                                                                                                                                                        
     Output layout: [orderkey_3, extendedprice, discount, suppkey, orderkey_3, $hashvalue_114, $hashvalue_114]                                                                                              
     Output partitioning: HASH [orderkey_3]                                                                                                                                                                 
     - Project => [extendedprice:double, discount:double, $hashvalue_114:bigint, orderkey_3:bigint, suppkey:bigint]                                                                                         
             $hashvalue_114 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey_3"), 0))                                                                                                 
         - TableScan[hive:hive:default:lineitem, originalConstraint = true] => [orderkey_3:bigint, suppkey:bigint, extendedprice:double, discount:double]                                                   
                 LAYOUT: hive                                                                                                                                                                               
                 orderkey_3 := HiveColumnHandle{clientId=hive, name=orderkey, hiveType=bigint, hiveColumnIndex=0, columnType=REGULAR}                                                                       
                 suppkey := HiveColumnHandle{clientId=hive, name=suppkey, hiveType=bigint, hiveColumnIndex=2, columnType=REGULAR}                                                                           
                 extendedprice := HiveColumnHandle{clientId=hive, name=extendedprice, hiveType=double, hiveColumnIndex=5, columnType=REGULAR}                                                               
                 discount := HiveColumnHandle{clientId=hive, name=discount, hiveType=double, hiveColumnIndex=6, columnType=REGULAR}
```

Peak memory usage without this optimization:
`527M`
Peak memory usage with optimization:
`449M`

File: presto-main/src/main/java/com/facebook/presto/operator/FilterAndProjectOperator.java [hunk 1]
```diff
@@ -36,17 +38,19 @@
     private final PageBuilder pageBuilder;
     private final PageProcessor processor;
     private final String processingOptimization;
+    private final boolean hasFilter;
     private Page currentPage;
     private int currentPosition;
     private boolean finishing;
 
-    public FilterAndProjectOperator(OperatorContext operatorContext, Iterable<? extends Type> types, PageProcessor processor)
+    public FilterAndProjectOperator(OperatorContext operatorContext, Iterable<? extends Type> types, PageProcessor processor, boolean hasFilter)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/FilterAndProjectOperator.java]: When non-columnar processing is better?

Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/FilterAndProjectOperator.java]: never mind - discussed offline

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java [hunk 2]
```diff
@@ -87,6 +89,8 @@
 public class PageProcessorCompiler
         implements BodyCompiler<PageProcessor>
 {
+    private static final RowExpression TRUE_PREDICATE = constant(true, BOOLEAN);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: extract change in compiler to separate commit

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java [hunk 3]
```diff
@@ -263,6 +300,77 @@ private static MethodDefinition generateProjectColumnarMethod(
     {
         Parameter session = arg("session", ConnectorSession.class);
         Parameter page = arg("page", Page.class);
+        Parameter pageBuilder = arg("pageBuilder", PageBuilder.class);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: too much copy paste

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: done

**Comment by sopel39** [comment]
I've seen some `testLargeIn` failures. Although I don't know if their are related to this patch.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java (hunk 3 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: I can try to extract common parts somehow, but I am not sure if it will be more elegant.

**Comment by sopel39** [comment]
It seems that there is also a problem with symbol pruning in more complex queries (e.g: in `Q5` `Projection` before `InnerJoin` doesn't look to be required). We will investigate that too.

File: presto-main/src/main/java/com/facebook/presto/operator/FilterAndProjectOperator.java [hunk 4]
```diff
@@ -94,6 +98,13 @@ public final void addInput(Page page)
     public final Page getOutput()
     {
         if (!pageBuilder.isFull() && currentPage != null) {
+            if (!hasFilter && processingOptimization.equals(FeaturesConfig.ProcessingOptimization.DISABLED)) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/FilterAndProjectOperator.java]: done

**Comment by fmeiser** [comment]
@sopel39 What is the reason for columnar processing only without filters is possible?
Background of the question: If it is only a property of some of the filters, a generic implementation with a method on the filter would be possible.

**Comment by sopel39** [comment]
@fmeiser columnar processing is possible with and without filter. With filter though, you need to first compute indexes of selected rows (one additional pass). Therefore it might be more efficient to use non-columnar processing.

**Comment by fmeiser** [comment]
@sopel39 I ran `PageProcessorBenchmark` and in all cases `columnOriented` is faster than `rowOriented`.

**Comment by sopel39** [comment]
@fmeiser did you use filter condition?

**Comment by fmeiser** [comment]
Yes, that is the default in current version of ''PageProcessorBenchmark''

**Comment by dain** [comment]
There is a typo in the first commit message `cas`

**Comment by sopel39** [comment]
refactored, please review again @losipiuk

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java (hunk 3 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java [hunk 5]
```diff
@@ -99,15 +104,18 @@ public PageProcessorCompiler(Metadata metadata)
...
         ImmutableList.Builder<MethodDefinition> projectDictionaryMethods = ImmutableList.builder();
         for (int i = 0; i < projections.size(); i++) {
             MethodDefinition project = generateProjectMethod(classDefinition, callSiteBinder, cachedInstanceBinder, "project_" + i, projections.get(i));
-            MethodDefinition projectColumnar = generateProjectColumnarMethod(classDefinition, callSiteBinder, "projectColumnar_" + i, projections.get(i), project);
-            MethodDefinition projectRLE = generateProjectRLEMethod(classDefinition, "projectRLE_" + i, projections.get(i), project, projectColumnar);
-            MethodDefinition projectDictionary = generateProjectDictionaryMethod(classDefinition, "projectDictionary_" + i, projections.get(i), project, projectColumnar, projectRLE);
+            MethodDefinition projectColumnar = generateProjectColumnarMethod(classDefinition, callSiteBinder, "projectColumnar_" + i, projections.get(i), project, hasFilter);
+            MethodDefinition projectColumnarWithSelectedPositions = generateProjectColumnarMethod(classDefinition, callSiteBinder, "projectColumnarWithSelectedPositions_" + i, projections.get(i), project, true);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: add a comment why you generate method with `true`

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java [hunk 6]
```diff
@@ -99,15 +104,18 @@ public PageProcessorCompiler(Metadata metadata)
     @Override
     public void generateMethods(ClassDefinition classDefinition, CallSiteBinder callSiteBinder, RowExpression filter, List<RowExpression> projections)
     {
+        boolean hasFilter = !filter.equals(TRUE_PREDICATE);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: shouldn't you apply change of same shape to `GenericPageProcessor`.
So the compiled and non-compiled one correspond to each other.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java [hunk 7]
```diff
@@ -289,11 +313,21 @@ private static MethodDefinition generateProjectColumnarMethod(
 
         Variable positionCount = scope.declareVariable("positionCount", body, page.invoke("getPositionCount", int.class));
         Variable position = scope.declareVariable("position", body, constantInt(0));
-        Variable cardinality = scope.declareVariable("cardinality", body, selectedPositions.length());
         Variable outputBlock = scope.declareVariable(Block.class, "outputBlock");
         Variable blockBuilder = scope.declareVariable("blockBuilder", body, pageBuilder.invoke("getBlockBuilder", BlockBuilder.class, projectionIndex));
         Variable type = scope.declareVariable("type", body, pageBuilder.invoke("getType", Type.class, projectionIndex));
 
+        Variable cardinality;
+        BytecodeExpression loopCounter;
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: I would rather call current for control variable `position` sth like `loopCounter` and use `position` as thing you are calling `loopCounter`.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java [hunk 8]
```diff
@@ -703,16 +737,16 @@ private static BytecodeBlock getBytecodeFilterOnDictionary(
...
-            .ifTrue(selectedDictionaryPositions.set(filterResult))
-            .ifFalse(new BytecodeBlock()
-                .append(new ForLoop()
-                            .initialize(position.set(constantInt(0)))
-                            .condition(lessThan(position, dictionaryPositionCount))
-                            .update(position.increment())
-                            .body(selectedDictionaryPositions.setElement(position, invokeFilter(thisVariable, session, singletonList(dictionary), position))))
-                .append(thisVariable.setField("inputFilterDictionary", dictionary))
-                .append(thisVariable.setField("filterResult", selectedDictionaryPositions))));
+                .condition(equal(dictionary, inputFilterDictionary))
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: revert

File: presto-main/src/main/java/com/facebook/presto/operator/FilterAndProjectOperator.java [hunk 9]
```diff
@@ -40,12 +42,20 @@
...
     private boolean finishing;
 
-    public FilterAndProjectOperator(OperatorContext operatorContext, Iterable<? extends Type> types, PageProcessor processor)
+    public FilterAndProjectOperator(OperatorContext operatorContext, Iterable<? extends Type> types, PageProcessor processor, boolean hasFilter)
     {
         this.processor = requireNonNull(processor, "processor is null");
         this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
         this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
-        this.processingOptimization = getProcessingOptimization(operatorContext.getSession());
+
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/FilterAndProjectOperator.java]: extract local `getProcessingOptimization` method

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java (hunk 6 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: `GenericPageProcessor`doesn't have all the optimizations from `PageProcessorCompiler` (e.g. does not check if projection is identity of constant. Look at `GenericPageProcessor#projectColumnar`

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java (hunk 5 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: added TODO

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java (hunk 7 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/FilterAndProjectOperator.java (hunk 9 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/FilterAndProjectOperator.java]: done

**Comment by sopel39** [comment]
applied comments

**Comment by sopel39** [comment]
FYI @martint this makes column copying for projections that don't have filter costless.

**Comment by dain** [comment]
@losipiuk if your changes are addressed, can you remove your changed requested flag?

**Comment by losipiuk** [comment]
Done

**Comment by dain** [comment]
Instead of this smaller change, I think we should switch to always using columnar, and only use the row oriented filter/project when we have a row oriented table scan.  As @fmeiser pointed out, the column oriented version is always faster (except for lazy row oriented datasources), and the reason we haven't flipped the default is the current code does not handle the expansion of pages in the columnar processor.  In the page processor we check the page size after each row is processes, and when it is the right size, we stop processing.

For columnar processing, I propose we do something similar to the row processor.  First we change the columnar processor to run over a fixed number of positions, say 1024, for each call.  Then after each column is projected we check the size, and if it is too big, we permanently reduce the number by half for that operator.  That way we only reduce when we find a problem, and at the worse case we end up processing a single 1024 segment twice.

What do you think?

**Comment by sopel39** [comment]
@dain I think this PR was more about eliminating unnecessary copying of columns so that channel shuffling projections become no-op and for projections without filter you only have to compute new columns. I planned that after that we could improve `PageProcessor` further. I am also fine with implementing your proposal in this PR if we are sure that columnar will always be faster. Did you actually run into cases where columnar vs row projection with filter had a significant impact on whole query? 

Are you sure that columnar processing will always perform better than row processing? Especially for small number of columns and simple filters building selection array might be a considerable overhead. I will re-check `PageProcessor` benchmarks.

Additionally, please have a read: http://15721.courses.cs.cmu.edu/spring2016/papers/p5-sompolski.pdf. Especially section 3 (`CASE STUDY: SELECT` and `figure 2`). In that section they evaluate columnar and row oriented methods for building selection array. Presto columnar processing corresponds to `Comp., if(1 && 2 && 3){} (lazy)`. It seems that depending on selectivity and predicates you should dynamically select optimal processing method. According to the paper, `Vectorized, non−branching` could improve filtering performance 2x for ~50-60% selectivity.  Additionally, running the least selective condition first so that latter conditions are not evaluated at all could also have a big performance impact.

**Comment by sopel39** [comment]
>For columnar processing, I propose we do something similar to the row processor. First we change the columnar processor to run over a fixed number of positions, say 1024, for each call. Then after each column is projected we check the size, and if it is too big, we permanently reduce the number by half for that operator. That way we only reduce when we find a problem, and at the worse case we end up processing a single 1024 segment twice.

@dain You would like proper block extension regardless of default columnar processing, right? What if block builder accepts 1025 rows? Then we will end up with chink size of 1? Should we care that much if a page exceeds size limit with few positions? Maybe we should just give up on last 1024 positions and add them to next page (page should accept at least one 1024 chunk)?

**Comment by dain** [comment]
> @dain You would like proper block extension regardless of default columnar processing, right? What if block builder accepts 1025 rows? Then we will end up with chink size of 1? Should we care that much if a page exceeds size limit with few positions? Maybe we should just give up on last 1024 positions and add them to next page (page builder should accept at least one 1024 chunk)?

No idea. We'll have to experiment to see what works best.

**Comment by dain** [comment]
> @dain I think this PR was more about eliminating unnecessary copying of columns so that channel shuffling projections become no-op and for projections without filter you only have to compute new columns. I planned that after that we could improve PageProcessor further. I am also fine with implementing your proposal in this PR if we are sure that columnar will always be faster. Did you actually run into cases where columnar vs row projection with filter had a significant impact on whole query?

Yes. We added this feature for performance improvements in an application.  I don't have the numbers right now, but IIRC there was a boost in the conversion to columnar, and a huge boost when you start processing dictionary columns.

> Are you sure that columnar processing will always perform better than row processing? Especially for small number of columns and simple filters building selection array might be a considerable overhead. I will re-check PageProcessor benchmarks.

Please do.  In general, I want to remove the manual switch, so you find cases that need the other form, we will need to build a heuristic to manage the switch over.

> Additionally, please have a read: http://15721.courses.cs.cmu.edu/spring2016/papers/p5-sompolski.pdf. Especially section 3 (CASE STUDY: SELECT and figure 2). In that section they evaluate columnar and row oriented methods for building selection array. Presto columnar processing corresponds to Comp., if(1 && 2 && 3){} (lazy). It seems that depending on selectivity and predicates you should dynamically select optimal processing method. According to the paper, Vectorized, non−branching could improve filtering performance 2x for ~50-60% selectivity. Additionally, running the least selective condition first so that latter conditions are not evaluated could also have a big performance impact.

In the past we tried a bunch of these kinds of tricks but saw a general boost.

**Comment by dain** [comment]
How much of this PR do we need if we make this change?

**Comment by sopel39** [comment]
So I see following tasks to be done:
* improve page building for columnar processing
* do benchmarks
* decide on default columnar processing

I think this PR has still value since projection without filter case will have a special case anyway. What do you think @dain?

**Comment by dain** [comment]
I only looked at the first commit, but the second one looks like something we might want in general.  I'll read it in detail and get back to you.

**Comment by sopel39** [comment]
rebased on master. This is prerequisite PR to following next steps:
- improve page building for columnar processing
- do benchmarks
- decide on default columnar processing
If you had a change to finish review @dain I could move forward with other steps.

**Comment by sopel39** [comment]
rebased & updated @dain

**Comment by sopel39** [comment]
rebased @dain

**Comment by dain** [comment]
I've started to dig into this problem in earnest, and have come to the conclusion that I would like to solve the general memory problem preventing us from turning on columnar all the time.  Additionally, the existing code is way to complex and is preventing us from quickly adding new features (e.g., optimizing filtering when there is no filter, convert constant projection into RLE block creation, etc).

The good news is after a few hours playing with this code, I think I have found an approach that is much simpler.  Instead of bytecode generating an entire PageProcess for columnar, we only need to generate the performance critical processors for a single expression over a page.  The rest of the code which manages special blocks, memory management and caches, can all be in plain old java.  I expect to have a PR in a day or two.

**Comment by sopel39** [comment]
@dain thanks for information. Please mention me in the PR so I could see how you implemented the changes.

**Comment by sopel39** [comment]
@dain have you managed to publish PR?

**Comment by dain** [comment]
@sopel39 I have one more test to write and then I will publish it (after I write a quick fix to a coordinator scaling problem)

**Comment by dain** [comment]
Sorry, this took so long to get up (last two weeks were super busy).  

The new page processor code can be found here: https://github.com/prestodb/presto/pull/7515
